text/0045-docker-proxy.md

```diff
@@ -0,0 +1,167 @@
+  The setting is site-specific for your corporate proxy server. For example,
+  `"httpProxy": "http://proxy.foocorp.com:8080",`
+* `httpsProxy` sets the proxy and port for HTTPS (TLS) requests. For example,
+  `"httpsProxy": "http://proxy.foocorp.com:4443",`
+* `ftpProxy` sets the proxy for FTP file transfer (this is not common)
+* `noProxy` sets a comma-separated list of networks and hosts that are in
+  the internal network. For example,
+  `"noProxy": "192.168.0.0/16,172.16.0.0/12,10.0.0.0,localhost,127.0.0.0/8,::1,foocorp.com",`
+
+  Make sure you include Docker Compose internal network(s), which by default
+   are within `172.16.0.0/12` (use `netstat -rn` to list your networks)
+* Do not set `http_proxy` , `https_proxy` , and `no_proxy` or their upper
+  case equivalents. This method has been replaced with the `config.json`
+  settings.
+* Proxy support should be briefly mentioned in the README.md file or similar
+  file, along with the proxy variable(s) needed, such as `httpProxy`
+
+The following is a generic example of `$HOME/.docker/config.json` . The
+actual contents are site specific:
+
+```
+{
+ "proxies":
+ {
+   "default":
+   {
+     "httpProxy": "http://proxy.foocorp.com:8080",
+     "httpsProxy": "http://proxy.foocorp.com:4443",
+     "noProxy": "192.168.0.0/16,172.16.0.0/12,10.0.0.0,localhost,127.0.0.0/8,::1,foocorp.com"
+   }
+ }
+}
+````
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+All Docker files must be updated to conform as follows:
+* Commands `apt, apt-get, cargo, deb, go, npm, pip3, wget` and others require
+  no changes
+* `curl` requires no changes, but it is not usually installed. It needs to be
+  installed for containers that use it. For example,
+  `RUN apt-get update && apt-get install -yq curl `
+
+* `apt-key` : Instead of using `apt-key adv` to download an apt signing key,
+  use `curl` and `apt-key add -` . For example, instead of the following in a
+  `Dockerfile` (which no longer works):
+
+```
+RUN \
+ apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 \
+ --recv-keys 8AA7AF1F1091A5FD
+```
+
+Use the following, which works in proxy and non-proxy environments:
+
+```
+RUN \
+ apt-get update && apt-get install -yq curl && curl -sSL \
+ 'http://keyserver.ubuntu.com/pks/lookup?op=get&search=0x8AA7AF1F1091A5FD’\
+ | apt-key add -
+```
+
+
+# Conclusion
+[conclusion]: #conclusion
+
+These recommendations allow Docker and Docker compose use with Sawtooth
+software in a standard, concise manner that work in both a corporate proxy
+and non-proxy (or transparent proxy) environments.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+None
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+The current alternative used is hand-modifying Docker scripts each time a
+proxy is required. This is time consuming, expensive, and unnecessary.
+
+An earlier alternative of using `apt-key adv` and passing proxy variables on
```

Comment:
This paragraph needs changes. Xenial did not require `--keyserver-options` explicitly passing proxy env. `config.json` in `~/.docker/` is sufficient for Xenial. The concern is however with Bionic, where all Sawtooth Dockerfiles are moved to.

Reply:
I do not mention that `--keyserver-options` is required (one can also use environment variables, for example, in Xenial).. I do mention that `--keyserver-options` was used on line 36 (not lines 140-143).

It is good to know `config.json` is also supported in Xenial, as it makes this proposal to use `config.jason` stronger.